### PR TITLE
fix(tests): add intern 4 support

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes https://github.com/mozilla/fxa-content-server/issues/5677
Fixes https://github.com/mozilla/fxa-content-server/issues/5341
Fixes https://github.com/mozilla/fxa-content-server/issues/5228

### CODE REVIEW

File: tests/intern4.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const intern = require('intern').default;
+
+intern.configure({
```
Comment (vladikoff) [inline tests/intern4.js]: Need more config stuff here

**Comment by vladikoff** [comment]
Runs via

```
node tests/intern4.js 
```

cc @shane-tomlinson

**Comment by vladikoff** [comment]
@jrgm @vbudhram @philbooth something pls review 

Latest docs: https://theintern.io/docs.html#Intern/4/docs/docs%2Fconfiguration.md/environmentspecific-config

Draft Migration Guide: https://theintern.io/docs.html#Intern/4/docs/docs%2Fchanges_from_3.md

File: tests/intern.js [hunk 2]
```diff
@@ -1,102 +1,132 @@
...
+if (args.firefoxBinary) {
+  config.capabilities['moz:firefoxOptions'].binary = args.firefoxBinary; //eslint-disable-line camelcase
+}
 
-  return config;
-});
+intern.configure(config);
+intern.run()
+  .catch((e) => {
+    // This might not throw, BUG filed: https://github.com/theintern/intern/issues/868
```
Comment (vladikoff) [inline tests/intern.js]: https://github.com/theintern/intern/issues/868

File: tests/intern.js [hunk 3]
```diff
@@ -1,102 +1,133 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Comment (philbooth) [inline tests/intern.js]: Was it intentional to delete the licensing preamble? Any reason?

File: tests/server/flow-event.js [hunk 4]
```diff
@@ -430,48 +428,49 @@ define([
     },
 
     'call flowEvent without optional flow data': {
-      beforeEach () {
+      beforeEach() {
```
Comment (philbooth) [inline tests/server/flow-event.js]: In my usual grumpy manner, I'm going to point out that this massive diff would be easier to review if it didn't also contain a bunch of unrelated formatting changes like removing the space on all these `beforeEach` methods. No biggie obviously, just more to scroll through.

File: tests/server/flow-event.js [hunk 5]
```diff
@@ -800,318 +810,327 @@ define([
...
     },
 
     'call flowEvent with invalid navigationTiming': {
-      beforeEach () {
+      beforeEach() {
         flowMetricsValidateResult = true;
         setup({
           events: [
-            { offset: 1000, type: 'loaded' }
+            {offset: 1000, type: 'loaded'}
```
Comment (philbooth) [inline tests/server/flow-event.js]: ...and in all these object literals...

File: package.json [hunk 6]
```diff
@@ -11,14 +11,13 @@
...
-    "test-functional": "intern-runner config=tests/intern_functional",
-    "test-functional-oauth": "intern-runner config=tests/intern_functional_oauth",
-    "test-latest": "intern-runner config=tests/intern_functional_full fxaAuthRoot=https://latest.dev.lcip.org/auth/v1 fxaContentRoot=https://latest.dev.lcip.org/ fxaEmailRoot=http://restmail.net fxaOauthApp=https://123done-latest.dev.lcip.org/ fxaUntrustedOauthApp=https://321done-latest.dev.lcip.org/ fxaProduction=true fxaToken=https://token.dev.lcip.org/1.0/sync/1.5",
-    "test-sauce": "intern-runner config=tests/intern_sauce",
-    "test-server": "intern-client config=tests/intern_server",
-    "test-travis": "intern-runner config=tests/intern_travis"
+    "test": "node tests/intern.js --unit=true",
+    "test-circle": "node tests/intern.js --suites=circle --fxaAuthRoot=https://fxaci.dev.lcip.org/auth --fxaEmailRoot=http://restmail.net --fxaOauthApp=https://oauth-fxaci.dev.lcip.org --fxaUntrustedOauthApp=https://321done-fxaci.dev.lcip.org --fxaProduction=true --bailAfterFirstFailure=true",
+    "test-functional": "node tests/intern.js",
+    "test-functional-oauth": "node tests/intern.js -- --suites=oauth",
```
Comment (philbooth) [inline package.json]: How come this line has the extra `--` but the others don't? Is it required?

Comment (philbooth) [reply package.json]: Actually, it seems like the extra `--` is wrong.

If I run `npm run test-functional-oauth` with things as they are here, I get the full functional suite. But if I run `node tests/intern --suites=oauth` manually, I get just the OAuth functional suite.

File: tests/functional/complete_sign_up.js [hunk 7]
```diff
@@ -122,20 +117,19 @@ define([
...
+  beforeEach: function () {
+    email = TestHelpers.createEmail();
+  },
+  tests: {
     'open expired email verification link': function () {
       var verificationLink;
 
       return this.remote
-        // Sign up and obtain a verification link
+      // Sign up and obtain a verification link
```
Comment (philbooth) [inline tests/functional/complete_sign_up.js]: Another casualty of the war against unrelated whitespace?

File: tests/functional/connect_another_device.js [hunk 8]
```diff
@@ -156,9 +152,9 @@ define([
         .then(testUrlInclude(CONNECT_ANOTHER_DEVICE_ENTRYPOINT));
     },
 
-    'signin Fx Desktop, verify same browser': function () {
+    'signin Fx Desktop, verify same browser ': function () {
       const forceUA = UA_STRINGS['desktop_firefox'];
-      const query = { forceUA };
+      const query = {  forceUA };
```
Comment (philbooth) [inline tests/functional/connect_another_device.js]: ...but then here we're inserting extra whitespace.

File: tests/functional/connect_another_device.js [hunk 9]
```diff
@@ -179,16 +175,17 @@ define([
...
 
       return this.remote
-        // preVerified: false causes the "verify account" email to be sent,
-        // that's used later to verify.
-        .then(createUser(signUpEmail, PASSWORD, { preVerified: false }))
-        .then(createUser(signInEmail, PASSWORD, { preVerified: true }))
+      // preVerified: false causes the "verify account" email to be sent,
+      // that's used later to verify.
+        .then(createUser(signUpEmail, PASSWORD, {preVerified: false}))
+        .then(createUser(signInEmail, PASSWORD, {preVerified: true}))
```
Comment (philbooth) [inline tests/functional/connect_another_device.js]: I thought these made more sense when they were indented to the same level as the code they're commenting.

File: tests/functional/fx_desktop_handshake.js [hunk 10]
```diff
@@ -241,7 +237,7 @@ define([
 
     'Sync signin page - no user signed into browser, user signed in locally': function () {
       return this.remote
-        // First, sign in the user to populate localStorage
+      // First, sign in the user to populate localStorage
         .then(openPage(SIGNIN_PAGE_URL, selectors.SIGNIN.HEADER, {
```
Comment (philbooth) [inline tests/functional/fx_desktop_handshake.js]: ...and these comments.

**Comment by philbooth** [comment]
Only one of my points is something that's actually broken: `npm run test-functional-oauth`. And the licensing preamble in `tests/intern.js` seems like it should be there too. So provisional r+ based on those 2 being addressed.

The whitespace stuff I'm not bothered about, it just tripped my pointless-code-review-effort switch (which is tripped pretty easily, you may have noticed).

File: package.json (hunk 6 above)
Comment (vladikoff) [reply package.json]: good find!

File: tests/functional/change_password.js [hunk 11]
```diff
@@ -1,74 +1,70 @@
+'use strict';
```
Comment (vbudhram) [inline tests/functional/change_password.js]: Noticed some test files have the `use strict` and some don't. Should it be put on all?

File: tests/functional/fx_desktop_handshake.js (hunk 10 above)
Comment (vbudhram) [reply tests/functional/fx_desktop_handshake.js]: I've seen this when using WebStorm, it formats things unexpectedly.

File: tests/server/routes/get-verify-email.js [hunk 12]
```diff
@@ -97,26 +90,30 @@ define([
...
         assert.equal(ravenMock.ravenMiddleware.captureError.callCount, 0);
 
-        req.query.something = 'else';
-
-        mockModule(mocks).process(req, res, () => {
+        res.redirect = () => {
+          // calling with `req.query.something` captures a message to Sentry
           assert.equal(logger.error.callCount, 0);
-          assert.equal(ravenMock.ravenMiddleware.captureMessage.callCount, 0);
+          assert.equal(ravenMock.ravenMiddleware.captureMessage.callCount, 1);
```
Comment (vbudhram) [inline tests/server/routes/get-verify-email.js]: I'm a little confused here, test says `no logs` but it checks that it logged.

File: tests/server/routes/get-favicon.js [hunk 13]
```diff
@@ -1,164 +1,158 @@
...
+
+/*eslint-disable sorting/sort-object-props*/
+registerSuite('routes/get-favicon', {
+  'route interface is correct': function () {
+    assert.isFunction(route);
+    assert.lengthOf(route, 0);
+  },
+
+  'initialise route': {
+    setup: function () {
```
Comment (vbudhram) [inline tests/server/routes/get-favicon.js]: Should this be `before`?

Comment (vladikoff) [reply tests/server/routes/get-favicon.js]: yesss!

File: tests/server/routes/get-verify-email.js (hunk 12 above)
Comment (vladikoff) [reply tests/server/routes/get-verify-email.js]: updated description, good find!

File: tests/functional/change_password.js (hunk 11 above)
Comment (vladikoff) [reply tests/functional/change_password.js]: cleaned these up!
